### PR TITLE
ASM-4988 Destroy interface from all vLANS

### PR DESCRIPTION
**Pull request by logankimmel**
Added ensure => absent to remove the interface from all
switch vLAN's

### CODE REVIEW

File: lib/puppet/provider/force10_interface/dell_ftos.rb [hunk 1]
```diff
@@ -17,4 +17,69 @@ def flush
...
+
+  def destroy
+    Puppet.debug("Interfaces to destroy: #{@ifaces_to_destroy}")
+    # translate fact key to switch command
+    iface_map = {
+        'tengigabit' => 'tengigabitethernet',
+        'fortygigabit' => 'fortyGigE',
+        'gigabit' => 'gigabitethernet'
+    }
+    # binding.pry
```
Comment (logankimmel) [inline lib/puppet/provider/force10_interface/dell_ftos.rb]: I'll remove this

File: lib/puppet/provider/force10_interface/dell_ftos.rb [hunk 2]
```diff
@@ -17,4 +17,69 @@ def flush
...
+        speed = iface_map[interface.split('_')[1]]
+        command_str = "no #{type} #{speed} #{@iface}"
+        transport.session.command(command_str)
+      end
+      transport.session.command('exit')
+    end
+  end
+
+  def create
+    #Does nothing right now
```
Comment (gavin-scott) [inline lib/puppet/provider/force10_interface/dell_ftos.rb]: How does this interact with the pre-existing configuration behavior that allowed setting e.g. mtu, portfast etc.? I assume that with `ensure = present` those values will still be set even though `create` does nothing? 

What about if a resource has `ensource = absent` and also specifies values for mtu, portfast, etc.? Is the  behavior that the VLANs are removed and mtu and portfast are set to the specified values?

**Comment by gavin-scott** [comment]
I'd love to see some unit tests on this. It looks like there is an existing unit test for this type that @asaenz added.

File: lib/puppet/provider/force10_interface/dell_ftos.rb (hunk 2 above)
Comment (logankimmel) [inline lib/puppet/provider/force10_interface/dell_ftos.rb]: @gavin-scott That is correct.  On `ensure => present`, the interface is created with the settings specified, just as before.  If you were to use `ensure => absent` witch other interface config settings as well.  The interface will still get configured with these settings but will not be on any vlan
The site.pp that I tested with was such:
```puppet
node default {
asm::mxl { '18':
  vlan_name => '',
  untagged_tengigabitethernet => '0/1',
  tagged_portchannel => '',
  require => 'Force10_interface[Tengigabitethernet0/1]',
}
force10_interface { 'Tengigabitethernet0/1':
  switchport => 'true',
  shutdown => 'false',
  portmode => 'hybrid',
  portfast => 'portfast',
  edge_port => 'edge-port',
  mtu => '12000',
  protocol => 'lldp',
  ensure => absent,
}
}
```

**Comment by logankimmel** [comment]
@gavin-scott this is G2G

File: lib/puppet_x/force10/model/switch.rb [hunk 3]
```diff
@@ -47,6 +47,12 @@ def skip_params_to_hash
     [ :snmp, :archive ]
   end
 
+  def all_vlans
+   # binding.pry
+    grp = PuppetX::Force10::Model.const_get('vlan'.capitalize).new(transport, facts, {:name => 'vlan'})
```
Comment (ripienaar) [inline lib/puppet_x/force10/model/switch.rb]: I am probably missing something obvious but why the ```const_get```? that tends to be used only when you have a variable class name?

**Comment by ripienaar** [comment]
We should probably do this in a 2nd PR but we need similar handling for ```force10_fcoemap``` and ```force10_zoneset```

File: lib/puppet_x/force10/model/switch.rb (hunk 3 above)
Comment (ripienaar) [inline lib/puppet_x/force10/model/switch.rb]: just to be clear, from what I can see the above is identical to:

```
grp = PuppetX::Force10::Model::Vlan.new(transport, facts, {:name => 'vlan'})
```

Should also probably kill the commented out ```binding.pry``` there

**Comment by logankimmel** [comment]
@gavin-scott @ripienaar  I added support for s-series by moving the vlan_information fact into the base possible_facts.  This way the same logic can be used for removing the port interface from the vlans.

Added tests for s_series parsing of interfaces to remove

**Comment by gavin-scott** [comment]
I think this is good to go, I'm going to go ahead and merge. @ripienaar note that @logankimmel refactored some of the facts to be more consistent between IOM and ToR -- it might simplify the switch config code a bit to adjust to that change at some point. He left the old facts in place so that nothing breaks.

**Pull request merged**